controllers/auth.js

```diff
@@ -9,6 +9,81 @@ const {
   USER_DOES_NOT_EXIST_ERROR,
 } = require("../constants/errorMessages");
 
+const googleAuthLogin = (req, res, next) => {
+  const { redirectURL, dev } = req.query;
+  if (dev === "true") {
+    return passport.authenticate("google", {
+      scope: ["email"],
+      state: redirectURL,
+    })(req, res, next);
+  } else {
+    return res.boom.unauthorized("User cannot be authenticated");
+  }
+};
+
+const googleAuthCallback = (req, res, next) => {
+  const rdsUiUrl = new URL(config.get("services.rdsUi.baseUrl"));
+  let authRedirectionUrl = rdsUiUrl;
+
+  if ("state" in req.query) {
+    try {
+      const redirectUrl = new URL(req.query.state);
+
+      if (`.${redirectUrl.hostname}`.endsWith(`.${rdsUiUrl.hostname}`)) {
+        // Matching *.realdevsquad.com
+        authRedirectionUrl = redirectUrl;
+      } else {
+        logger.error(`Malicious redirect URL provided URL: ${redirectUrl}, Will redirect to RDS`);
+      }
+    } catch (error) {
+      logger.error("Invalid redirect URL provided", error);
+    }
+  }
+  try {
+    return passport.authenticate("google", { session: false }, async (err, accessToken, user) => {
+      if (err) {
+        logger.error(err);
+        return res.boom.unauthorized("User cannot be authenticated");
+      }
+      const userData = {
+        email: user.emails[0].value,
+        created_at: Date.now(),
+        updated_at: null,
+      };
+
+      const userDataFromDB = await users.fetchUser({ email: userData.email });
+
+      if (userDataFromDB.userExists) {
+        if (userDataFromDB.user.roles.developer === true) {
+          const errorMessage = encodeURIComponent("Google login is restricted for developer role.");
+          return res.redirect(`${authRedirectionUrl}?error=${errorMessage}`);
+        }
+      }
+
+      const { userId, incompleteUserDetails } = await users.addOrUpdate(userData);
+
+      const token = authService.generateAuthToken({ userId });
+
+      const cookieOptions = {
+        domain: rdsUiUrl.hostname,
+        expires: new Date(Date.now() + config.get("userToken.ttl") * 1000),
+        httpOnly: true,
+        secure: true,
+        sameSite: "lax",
+      };
+
+      res.cookie(config.get("userToken.cookieName"), token, cookieOptions);
+
+      if (incompleteUserDetails) authRedirectionUrl = "https://my.realdevsquad.com/new-signup";
+
+      return res.redirect(authRedirectionUrl);
+    })(req, res, next);
```

Comment:
- let's move the logic to a different function instead of adding it inside the callback function of the `passport.authenticate` method.

Reply:
Will move to a different function.